src/runtime/opencl.cpp

```diff
@@ -1488,3 +1493,386 @@ WEAK halide_device_interface_t opencl_device_interface = {
 }  // namespace Internal
 }  // namespace Runtime
 }  // namespace Halide
+
+extern "C" {
+
+WEAK int halide_opencl_image_device_malloc(void *user_context, halide_buffer_t *buf) {
+    debug(user_context)
+        << "CL: halide_opencl_image_device_malloc (user_context: " << user_context
+        << ", buf: " << buf << ")\n";
+
+    ClContext ctx(user_context);
+    if (ctx.error_code != CL_SUCCESS) {
+        return ctx.error_code;
+    }
+
+    size_t size = buf->size_in_bytes();
+    halide_assert(user_context, size != 0);
+    if (buf->device) {
+        halide_assert(user_context, validate_device_pointer(user_context, buf, size));
+        return 0;
+    }
+
+    for (int i = 0; i < buf->dimensions; i++) {
+        halide_assert(user_context, buf->dim[i].stride >= 0);
+    }
+
+    debug(user_context) << "    allocating " << *buf << "\n";
+
+#ifdef DEBUG_RUNTIME
+    uint64_t t_before = halide_current_time_ns(user_context);
+#endif
+
+    device_handle *dev_handle = (device_handle *)malloc(sizeof(device_handle));
+    if (dev_handle == NULL) {
+        return CL_OUT_OF_HOST_MEMORY;
+    }
+
+    cl_image_format format;
+    cl_image_desc desc;
+
+    struct halide_type_t type = buf->type;
+    if (type.code == halide_type_int) {
+        if (type.bits == 8) {
+            format.image_channel_data_type = CL_SIGNED_INT8;
+        } else if (type.bits == 16) {
+            format.image_channel_data_type = CL_SIGNED_INT16;
+        } else if (type.bits == 32) {
+            format.image_channel_data_type = CL_SIGNED_INT32;
+        } else {
+            halide_assert(user_context, false && "unhandled int bit width for image");
+        }
+    } else if (type.code == halide_type_uint) {
+        if (type.bits == 8) {
+            format.image_channel_data_type = CL_UNSIGNED_INT8;
+        } else if (type.bits == 16) {
+            format.image_channel_data_type = CL_UNSIGNED_INT16;
+        } else if (type.bits == 32) {
+            format.image_channel_data_type = CL_UNSIGNED_INT32;
+        } else {
+            halide_assert(user_context, false && "unhandled uint bit width for image");
+        }
+    } else if (type.code == halide_type_float) {
+        if (type.bits == 16) {
+            format.image_channel_data_type = CL_HALF_FLOAT;
+        } else if (type.bits == 32) {
+            format.image_channel_data_type = CL_FLOAT;
+        } else {
+            halide_assert(user_context, false && "unhandled float bit width for image");
+        }
+    } else {
+        halide_assert(user_context, false && "unhandled data type for image");
+    }
+
+    int last_dim_size = buf->dim[buf->dimensions - 1].extent;
```

Comment:
`last_dim_size` is unused, which will generate a warning-treated-as-error on some compilers (notable xcode, see https://buildbot.halide-lang.org/master/#/builders/10/builds/41)